addons/project/models/project.py

```diff
@@ -945,18 +939,32 @@ def _compute_subtask_count(self):
 
     @api.onchange('company_id')
     def _onchange_task_company(self):
-        if self.project_id.company_id != self.company_id:
-            self.project_id = False
+        if self.effective_project_id.company_id != self.company_id:
+            self.effective_project_id = False
 
-    @api.depends('project_id.company_id')
+    @api.onchange('recurring_task')
+    def _onchange_recurring_task(self):
+        self.ensure_one()
+        if self.parent_id and self.recurring_task:
+            raise UserError(_("A sub-task cannot be a recurring task. Please consider making its parent task, a recurring task."))
+
+    @api.depends('effective_project_id.company_id')
     def _compute_company_id(self):
-        for task in self.filtered(lambda task: task.project_id):
-            task.company_id = task.project_id.company_id
+        for task in self.filtered(lambda task: task.effective_project_id):
+            task.company_id = task.effective_project_id.company_id
 
-    @api.depends('project_id')
+    @api.onchange('stage_id')
+    def _onchange_stage_id(self):
+        self.ensure_one()
+        if self.parent_id and not self.project_id and self.parent_id.stage_id != self.stage_id:
+            raise UserError(_("The sub-task stage cannot be different from the one of its parent."))
```

Comment:
Is this really functionally intended ? 

What about the parent task "Refactor Project" that remain in "Development" state during 5 months, and all the subtasks like "Refactor Subtasks" ;)

Reply:
The functional need is "We assume that a sub-task has to be in the same stage and project as its parent task EXCEPT if the project_id field is set".

I guess there will be a different workflow to do the "Refactor Project">"Refactor Subtasks" use case. As I understood, user will probably need to use the task dependency or milestones rather than Task>Subtask.